proto/structured.proto

```diff
@@ -0,0 +1,95 @@
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+syntax = "proto2";
+package cockroach.proto;
+option go_package = "proto";
+
+import "gogoproto/gogo.proto";
+import "cockroach/proto/api.proto";
+import "cockroach/proto/errors.proto";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+message Column {
+  enum ColumnType {
+    // option (gogoproto.goproto_enum_prefix) = false;
+    BYTES = 0;
+  }
+
+  optional string name = 1 [(gogoproto.nullable) = false];
+  optional ColumnType type = 2 [(gogoproto.nullable) = false];
+}
+
+message Index {
+  optional string name = 1 [(gogoproto.nullable) = false];
+  optional bool unique = 2 [(gogoproto.nullable) = false];
+}
+
+message TableSchema {
+  message IndexByName {
+    optional Index index = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+    // An ordered list of column names of which the index is comprised. Order
+    // matters!
+    repeated string column_names = 2 [(gogoproto.nullable) = false];
+  }
+  optional string name = 1 [(gogoproto.nullable) = false];
+  repeated Column columns = 2 [(gogoproto.nullable) = false];
+  // An ordered list of indexes included in the table. The first index is the
+  // primary key; it is required.
+  repeated Index indexes = 3 [(gogoproto.nullable) = false];
+}
+
+// A TableDescriptor represents a table and is stored in a structured metadata
+// key. The TableDescriptor has a globally-unique ID, while its member
+// {Column,Index}Descriptors have locally-unique IDs.
+message TableDescriptor {
+  message ColumnDescriptor {
+    optional uint32 id = 1 [(gogoproto.nullable) = false];
+    optional Column column = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  }
+  message IndexDescriptor {
+    optional uint32 id = 1 [(gogoproto.nullable) = false];
+    optional Index index = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+    // Each column_id refers to a column in the TableDescriptor's columns.
+    // Special care is taken to update this when deleting columns.
+    repeated uint32 column_ids = 3 [(gogoproto.nullable) = false];
+  }
+  optional uint32 id = 1 [(gogoproto.nullable) = false];
+  repeated ColumnDescriptor columns = 2 [(gogoproto.nullable) = false];
+  // next_column_id is used to ensure that deleted column ids are not reused
+  optional uint32 next_column_id = 3 [(gogoproto.nullable) = false];
+  repeated IndexDescriptor indexes = 4 [(gogoproto.nullable) = false];
+  // next_index_id is used to ensure that deleted index ids are not reused
+  optional uint32 next_index_id = 5 [(gogoproto.nullable) = false];
+}
+
+message CreateTableRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional TableSchema schema = 2 [(gogoproto.nullable) = false];
```

Comment:
Yeah, it's possible we do, for convenience. I've omitted that for now since the table name is exactly the suffix of the TableDescriptor's key. It's an easy addition, let's see how it shakes out.